SHS-6399: Basic admin bar and edit own content permissions for authors

dalin- opened the pull request:
# User Story
As a user with the Author role

I see
* That I'm logged in (and thus might have access to things that the general public cannot)
* ~~How to get to those things (e.g. edit my content) rather than just stumbling through the site and occasionally seeing an "Edit" tab.~~

## Backstory
  [SHS-6399](https://app.clickup.com/t/36718269/SHS-6399)

## Tour
https://www.loom.com/share/430cf64d57ab4894ae75dad6e74b3a7a

## Need Review By (Date)
November 11/5

## Urgency
medium

## Steps to Test

1. Log into a test site
2. Assign the Author role to IT Site Manager and remove the Site Manager role
3. Edit at least 1 node and make IT Site Manager the author 
5. Masquerade as IT Site Manager
6. ~~You should end up on `/my-content`~~
7. Confirm you can see the menu bar at the top ~~and the list of content that IT Site manager is an author of under My Content~~
8. ~~Confirm you can access and edit the content listed~~
9. Log out from the top menu

## Known Issues
* ~Tugboat is not building.~
* If you try to use the admin toolbar to click on your name, you will have both toolbars open at the same time.  We'll investigate this in a follow-up ticket. 
* ~need to write an update hook to give authors permission to see the toolbar~

---
- To see the specific tasks where the Asana app for GitHub is being used, see below:
  - https://app.asana.com/0/0/1211726079591699

dalin- committed:
feat(SHS-6399): shortcuts and view for low-level roles

dalin- committed:
fix(SHS-6399): odd button for my-content link

dalin- committed:
fix(SHS-6399): don't redirect if pwd reset is in progress

dalin- committed:
fix(SHS-6399): more robust default to shortcuts menu

dalin- committed:
fix(SHS-6399): avoid cluttering the shortcutset_users table

ahughes3 commented:
[Tugboat](https://www.tugboatqa.com) has finished building the preview for this pull request!

Links:
- https://pr2029-pkwfo7knosnt1hmlwma5okzwxxflwmb8.tugboatqa.com
- https://hs-colorful-pkwfo7knosnt1hmlwma5okzwxxflwmb8.tugboatqa.com
- https://hs-traditional-pkwfo7knosnt1hmlwma5okzwxxflwmb8.tugboatqa.com
- https://english-pkwfo7knosnt1hmlwma5okzwxxflwmb8.tugboatqa.com
- https://history-pkwfo7knosnt1hmlwma5okzwxxflwmb8.tugboatqa.com
- https://west-pkwfo7knosnt1hmlwma5okzwxxflwmb8.tugboatqa.com


Dashboard:
- https://dashboard.tugboatqa.com/68f7ec079860dae97bc59f6c

dalin- committed:
feat(SHS-6399): give authors permission to see the toolbar

ahughes3 commented:
[Tugboat](https://www.tugboatqa.com) has finished building the preview for this pull request!

Links:
- https://pr2029-pkwfo7knosnt1hmlwma5okzwxxflwmb8.tugboatqa.com
- https://hs-colorful-pkwfo7knosnt1hmlwma5okzwxxflwmb8.tugboatqa.com
- https://hs-traditional-pkwfo7knosnt1hmlwma5okzwxxflwmb8.tugboatqa.com
- https://english-pkwfo7knosnt1hmlwma5okzwxxflwmb8.tugboatqa.com
- https://history-pkwfo7knosnt1hmlwma5okzwxxflwmb8.tugboatqa.com
- https://west-pkwfo7knosnt1hmlwma5okzwxxflwmb8.tugboatqa.com


Dashboard:
- https://dashboard.tugboatqa.com/68f7ec079860dae97bc59f6c

cienvaras commented on config/default/block.block.humsci_traditional_userlogin.yml:
```diff
@@ -0,0 +1,33 @@
+uuid: e777ce4e-d402-4de3-8474-ebb4482f3f4b
```
Is there a reason for including this config file? I think it's unrelated with the feature being added on this PR. If that's the case, please remove it.

cienvaras reviewed (commented):
@dalin- Looks good! I just left a question in the code about one of the exported confg files.

dalin- committed:
fix: delete accidentally added file

cienvaras reviewed (approved):
@dalin- LGTM, thanks for the fix.

@ahughes3 @Jnwibowo7321 Ready for you to review

ahughes3 commented:
[Tugboat](https://www.tugboatqa.com) has finished building the preview for this pull request!

Links:
- https://pr2029-pkwfo7knosnt1hmlwma5okzwxxflwmb8.tugboatqa.com
- https://hs-colorful-pkwfo7knosnt1hmlwma5okzwxxflwmb8.tugboatqa.com
- https://hs-traditional-pkwfo7knosnt1hmlwma5okzwxxflwmb8.tugboatqa.com
- https://english-pkwfo7knosnt1hmlwma5okzwxxflwmb8.tugboatqa.com
- https://history-pkwfo7knosnt1hmlwma5okzwxxflwmb8.tugboatqa.com
- https://west-pkwfo7knosnt1hmlwma5okzwxxflwmb8.tugboatqa.com


Dashboard:
- https://dashboard.tugboatqa.com/68f7ec079860dae97bc59f6c

dalin- committed:
Remove "Preparer" from low-level roles.

dalin- committed:
Remove "Preparer" from low-level roles.

Jnwibowo7321 reviewed (approved):
LGTM on Tugboat!

ahughes3 commented:
When testing on tugboat the My Content page can have content listed that the author cannot edit. It looks like the list is showing all content that the user is an author of but the permissions aren't set up for an author to edit all content even if they are the author. This can be tested by assigning an author content that is not a Person node. Before we merge we will need to either update the permissions so that authors can edit any content they are authors of or update the my content list to only be Person nodes/content that the author can edit.

dalin- committed:
feat(SHS-6399): give authors permission to edit all of their own content

dalin- committed:
chore(SHS-6427): linting issues

dalin- committed:
Merge remote-tracking branch 'origin/11.20.2-release' into SHS-6399_Implement-Basic-admin-bar-for-low-level-editors

* origin/11.20.2-release: (33 commits)
  Add release tag conditional to recent Node step additions in release workflow (#2058)
  SHS-6432: Fix | Sass tests are failing from within the theme (#2047)
  Automated Dependency Updates 20251114 (#2056)
  Fix npm ci failures by updating theme lockfile for Node v24.11.1+ (#2057)
  HSD8-1797: Remove machine_name column from Course Tag Form (#2055)
  HSD8-1797: Adjust Course Tag Table labels and Add Translated Tag Column (#2054)
  11.20.2
  Update profile version.
  Add node install and npm install steps to release workflow.
  Updated profile version.
  SHS-6460: Characters not displaying in Color Band when limited width (#2048)
  Revert "SHS-6346: Lazy load videos (#2004)"
  HSD8-1789: Grant hs_basic_block edit/create permissions via update hook (#2050)
  HSD8-1799: Grant hs_basic_block edit permission via update hook (#2049)
  SHS-6411: For mailing address, allow all countries, and make all subfields optional (#2046)
  SHS-6346: Lazy load videos (#2004)
  SHS-6439: Implement | Heading solution for accordions (#2040)
  Automated Dependency Updates 20251107 (#2045)
  SHS-6449: Orphaned pages | Add filter/column for "Published" (#2041)
  SHS-6383: Add and enable Edit Media Modal module to allow editing media used in CKEditor (#1980)
  ...

cienvaras commented:
@Jnwibowo7321 @ahughes3 [Requested changes](https://app.clickup.com/t/36718269/SHS-6399?comment=90130191468499) done, ready for review again.

joegl commented:
I think this one is good to go but I just want to confirm with @ahughes3 that the author role should have have "Edit own" permissions for every single content type (including private pages for example). I know when we talked about this it wasn't clear if that would be the case or not.

dalin- committed:
feat(SHS-6399): only show the shortcuts for high-level roles

dalin- commented:
Still not sure why these update hooks are not running on Tugboat.  They run fine locally.  This isn't going to make it into this release.  We'll have to bump this to the next one.

dalin- committed:
feat: don't go to the my-content list on login until Albert decides what a better option might be

ahughes3 reviewed (approved):
looks good to me!

joegl commented on docroot/modules/humsci/hs_dashboard/hs_dashboard.install:
```diff
@@ -14,3 +14,36 @@ function hs_dashboard_update_10000() {
   }
 }
+
+/**
```
Update hooks look good but they probably belong in the `su_humsci_profile.install` file since they're more general and not specific to the `hs_dashboard` module functionality.

joegl reviewed (changes requested):
This LGTM however I don't think the database update and hook logic fits the use-case of the `hs_dashboard` module, and should be moved into the `hs_admin` module where shortcut set logic has been added in the past.

Alternatively, the best approach would be an `hs_shortcuts` module where we strip all shortcut logic out of `su_humsci_profile` and `hs_admin` but I don't think that's worth the refactor or time required right now.

dalin- committed:
chore(SHS-6399): moved update hooks to a more logical home

dalin- commented:
Hi all, 
It looks like the last comment here was months ago during my vacation and it dropped off my radar.  Sorry about that. 

The change is now complete.  

It looks like this has been reviewed by everyone, so I think this is ready to merge in?

joegl commented:
@ahughes3 Before I review (and likely merge) I wanted to double-check with you on this one.

ahughes3 commented:
@joegl @dalin- Can we get the Tugboat preview working again? I would like to review before approving, thanks!

dalin- commented:
@joegl 

Any idea why my Tugboat won’t build?  Same result if I try with/without a base, or rebuild/refresh/reset
https://dashboard.tugboatqa.com/69a89e93c97975c5da68df27

It fails pulling the db from acquia.

joegl committed:
Merge branch '11.25.1-release' into SHS-6399_Implement-Basic-admin-bar-for-low-level-editors

joegl commented:
> @joegl
> 
> Any idea why my Tugboat won’t build?  Same result if I try with/without a base, or rebuild/refresh/reset https://dashboard.tugboatqa.com/69a89e93c97975c5da68df27
> 
> It fails pulling the db from acquia.

Yea the methodology for pulling in the site databases had to change after the ACN upgrade and that didn't make it in to this PR. I merged the latest code and tugboat should be functional now.

ahughes3 commented:
[Tugboat](https://www.tugboatqa.com) has finished building the preview for this pull request!

Links:
- https://pr2029-kgx1z5t67ojkq57lrxtyftngoixeh4b9.tugboatqa.com
- https://hs-colorful-kgx1z5t67ojkq57lrxtyftngoixeh4b9.tugboatqa.com
- https://hs-traditional-kgx1z5t67ojkq57lrxtyftngoixeh4b9.tugboatqa.com
- https://english-kgx1z5t67ojkq57lrxtyftngoixeh4b9.tugboatqa.com
- https://history-kgx1z5t67ojkq57lrxtyftngoixeh4b9.tugboatqa.com
- https://west-kgx1z5t67ojkq57lrxtyftngoixeh4b9.tugboatqa.com


Dashboard:
- https://dashboard.tugboatqa.com/69a8abc619638a6039091002

Visual Diffs:
- https://dashboard.tugboatqa.com/69a8abc619638a6039091002/visualdiffs

dalin- commented:
@ahughes3 over to you to review.

ahughes3 commented:
@joegl @dalin- I reviewed this and it is working as expected

Jnwibowo7321 commented:
LGTM! Good for merging @joegl

joegl commented on config/default/user.role.authenticated.yml:
```diff
@@ -25,7 +24,6 @@ is_admin: false
 permissions:
   - 'access content'
   - 'access fontawesome additional settings'
-  - 'access shortcuts'
```
The `access shortcuts` permissions is getting removed from all authenticated users and added to individual roles. The `author` role is not getting the permission added, however, it is the only low-level role included in the login hook to get the `low-level` shortcuts. Are they still able to access the shortcuts they're getting assigned?

joegl commented on docroot/modules/humsci/hs_dashboard/hs_dashboard.module:
```diff
@@ -90,3 +102,52 @@ function hs_dashboard_entity_operation(EntityInterface $entity) {
+    }
+  }
+  // All other roles get the default shortcuts.
+  $shortcut_set_storage->unassignUser($account);
```
Won't this unset the shortcut set just assigned to the low-level-user role? This should probably be in an else condition.

joegl commented on docroot/modules/humsci/hs_dashboard/hs_dashboard.module:
```diff
@@ -90,3 +102,52 @@ function hs_dashboard_entity_operation(EntityInterface $entity) {
+ * Implements hook_user_login().
+ */
+function hs_dashboard_user_login(UserInterface $account) {
+  $shortcut_set_storage = Drupal::entityTypeManager()->getStorage('shortcut_set');
```
```suggestion
  $shortcut_set_storage = \Drupal::entityTypeManager()->getStorage('shortcut_set');
```

joegl commented on docroot/modules/humsci/hs_dashboard/hs_dashboard.module:
```diff
@@ -90,3 +102,52 @@ function hs_dashboard_entity_operation(EntityInterface $entity) {
+/**
+ * Implements hook_user_login().
+ */
+function hs_dashboard_user_login(UserInterface $account) {
```
Instead of using a `hook_user_login()` hook for this, I think the `hook_shortcut_default_set` hook is a better approach: https://api.drupal.org/api/drupal/core%21modules%21shortcut%21shortcut.api.php/function/hook_shortcut_default_set/10

> This hook allows modules to define default shortcut sets for a particular user that differ from the site-wide default (for example, a module may want to define default shortcuts on a per-role basis).
> The default shortcut set is used only when the user does not have any other shortcut set explicitly assigned to them.

Instead of assigning a specific shortcut set, you can just change what the default shortcut set is for the user. It also will evaluate constantly, and won't require a user login or database update hook.

joegl commented on docroot/modules/humsci/hs_admin/hs_admin.install:
```diff
@@ -49,3 +49,36 @@ function hs_admin_update_10001() {
+}
+
+/**
+ * Revoke permission to use shortcuts for everyone except high roles.
```
Redundant comment, but isn't removing the access shortcuts permission from low-level roles and then assigning them a shortcut set prevent them from using that shortcut set?

dalin- committed:
Merge remote-tracking branch 'origin/11.26.2-release' into SHS-6399_Implement-Basic-admin-bar-for-low-level-editors

* origin/11.26.2-release:
  Automated Dependency Updates 20260313 (#2155)
  11.26.2
  HSD8-1849: Restore environment-specific config locking and prefixing logic (#2152)
  SHS-6588: Padding removed on Spotlight text in Full Width (#2147)
  SHS-6548: Get DDev working well with Acquia Cloud next, and other modern versions (#2148)
  11.26.1
  Increment version number for release.
  SHS-6484: Remove blazy from codebase (#2135)

dalin- committed:
chore(SHS-6399): removed the parts of this PR that we are postponing

dalin- committed:
chore(SHS-6399): dynamically give authors access to the toolbar

dalin- commented:
@joegl 

TL;DR  Good finds.   Fixed.  Can you take another look?

This PR evolved a lot between when it was first created and today.  At first we were working with multiple roles, and giving them both access to the admin toolbar, and shortcuts.  But now we've paired it back to just giving authors access to the toolbar, and granting them access to edit anything that they create.  I've removed all the extra code.

Can you take another look?

joegl commented:
LGTM. I'm going to fix the conflict and then merge.

joegl committed:
Merge branch '11.26.2-release' into SHS-6399_Implement-Basic-admin-bar-for-low-level-editors
